### ISSUE TITLE
A second argument in Promise.all([...]) that returns a Record breaks the return signature

### ISSUE THREAD

**Issue by dannycochran** [microsoft/TypeScript#34924]
**TypeScript Version:** 

3.7.2

**Search Terms:**

Promise.all, Record

**Code**

```typescript
interface MoviesResponse {
    movies: [];
}

type CharactersResponse = Record<number, any>;

const getMovies = async (): Promise<MoviesResponse> => ({
    movies: [],
})

const getCharacters = async (): Promise<CharactersResponse> => ({
});

(async () => {
    const [{ movies }, characters] = await Promise.all([getMovies(), getCharacters()]);
    console.log(movies, characters);
})();
```

**Expected behavior:**

The first variable in the deconstructed array should be correctly typed as `MoviesResponse`, rather than `CharactersResponse`. This worked in 3.6.4.

**Actual behavior:**

Both variables in the deconstructed array are inferred to be `CharactersResponse`. So when I try to deconstruct the first argument doing `{ movies }`, it's an invalid destructuring.

Interestingly, if we update `CharactersResponse` to not be a map type, the problem goes away, e.g:

```
interface CharactersResponse {
    characters: any[];
}
const getCharacters = async (): Promise<CharactersResponse> => ({
    characters: [],
});
```

**Playground Link:**

http://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=17&pc=6#code/JYOwLgpgTgZghgYwgAgLIHsBuwIGcBKeADuiLigN4BQytyAtljrgFzIDaAugNxUC+VKmACeRFAGEAFnCiJIUAsVLlkAXmSEE6KABMAPCACu9AEbQANMjghhAPl5UtZMMgDmEMBmx41V3MJAEZAAKAEo2AAUodHpgcj0vZkJcEjIIWzUM4Oo6BiY8Ni5zflDBJ1wXdzApGTloXF84f0CQ8OQomLiIPRrZBHlFFOV0zJDqPlCHYKaAoLDRnLpyl3YKPO8GvksEaT6BzkaAdzhgFw7Y8gA6OAAbG+D2KsS8MMsq3rqFMM5JmiXldA3CCXG7oVzBRgbba7T64X4TMLcIA

**Related Issues:**

None found.

**Comment by nmain** [comment]
> Both movies and characters are inferred to be CharactersResponse

But the playground you linked to types `movies` as `any` (with an error)?

**Comment by dannycochran** [comment]
> > Both movies and characters are inferred to be CharactersResponse
> 
> But the playground you linked to types `movies` as `any` (with an error)?

Sorry, you're right. It's that the first argument coming back from `Promise.all` is typed as `CharactersResponse`, so when I try to deconstruct it, it shows up as any. Will update my comment.

**Comment by nmain** [comment]
I see.  Once the example is corrected, this appears to work as you'd like it to in 3.6.3, but not 3.7.2

**Comment by orta** [comment]
Maybe fixed by https://github.com/microsoft/TypeScript/pull/33707

**Comment by AnyhowStep** [comment]
This will be fixed by #33707 
I just tested

**Comment by andrewbranch** [comment]
Actually, this is a specific reproduction of #34937. This one came first by some number of minutes or hours, but the other has more discussion, so I’m going to mark this one duplicate—nothing personal @dannycochran 😄
